### DISCUSSION TITLE
block clone / reflink space savings reporting on pool level vs. dataset level

### DISCUSSION THREAD

**Discussion by hanneskasp** [openzfs/zfs#16024] [General]
Hello,
I noticed, that reflink / bcloneratio space savings are reported on zpool level instead of dataset level. Is that expected or a bug? I'm asking, because reflink only works within a dataset, so I assume reporting on space savings could also be reported "per dataset". The use-case is charge-back "per tenant" where every "tenant" would have its own dataset while showing how much data reflink / block cloning saved him

```
zpool get all | grep -i clon
pool4  bcloneused                     684G                           -
pool4  bclonesaved                    1.07T                          -
pool4  bcloneratio                    2.60x                          -
pool4  feature@block_cloning          active                         local
```
no output for this
`zfs get all /pool4/linktest | grep -i clon`

What do you think?

Best  regards,
Hannes

**Comment by amotin** [comment]
Block Cloning ZFS feature is effectively a lighter version of dedup.  It works on a pool (actually individual VDEVs) level and knows nothing about datasets.  The only reasons it may be limited to dataset are either Linux kernel restrictions on some of APIs used by reflink, or limitations of ZFS encryption, which may have different keys for different datasets.  It is impossible to provide per-dataset statistics for Block Cloning.

**Reply by hanneskasp** [reply]
ah okay, with your explanation it makes sense that one cannot provide information per dataset. Thanks!

I use Rocky 9.3 with 5.14.0-362.18.1.el9_3.0.1.x86_64 and I get this error

```
cp --reflink=always randomfile /pool4/linktest2/link1
cp: failed to clone '/pool4/linktest2/link1' from 'randomfile': Invalid cross-device link

```
